src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -2573,11 +2573,10 @@ Node *PhiNode::Ideal(PhaseGVN *phase, bool can_reshape) {
         igvn->replace_node(this, result);
         igvn->set_type(result, result->bottom_type());
 
-        // now transform the new nodes, and return the mergemem
-        for (MergeMemStream mms(result); mms.next_non_empty(); ) {
-          Node* phi = mms.memory();
-          mms.set_memory(phase->transform(phi));
-        }
+        // We could immediately transform the new Phi nodes here, but that can
+        // result in creating an excessive number of new nodes within a single
+        // IGVN iteration. We have put the Phi nodes on the IGVN worklist, so
+        // they are transformed later on in any case.
         hook->destruct(igvn);
```

Comment:
Do you need this `hook` node to preserve `new_base` now that you are not calling `PhaseGVN::transform` anymore?

Reply:
Thanks for the comments! I have now cleaned up according to your suggestions and checked that it still works (tests and code inspection). I also investigated why the eager transformation was added in the first place, but was unable to find much information. It is from initial load, and looking back even further in the pre-initial-load history, I see that the code was introduced back in 2000 already. The commit message is not really helpful, unfortunately. My best guess is that eagerly transforming the Phis may perhaps result in a preferable ordering of idealizations that introduces less (temporary) nodes. The extremas for "target" and "target-old" gives an indication of this (see the plots in the PR description), but in practice it has no significant observable effect.

Reply:
Thanks for digging deeper! I almost expected that the history is probably lost about this decision.